Added formplayer overview for HQ devs (#29112)
dimagi/commcare-hq: docs/formplayer.rst

machine is back up. During a formplayer deploy, when we have to restart all formplayer processes, a rolling
restart doesn't help uptime, since for every individual user their machine's process will be down while it restarts.

Routing implications for adding and removing machines
-----------------------------------------------------

It's expensive to delete a user's sqlite sandbox, because rebuilding it requires requesting a full restore from
HQ, but it's always **safe** to delete it, because that rebuild from the source of truth will get the user back to
normal. This property makes removing machines a safe operation.
Similarly, adding new machines doesn't pose an issue because the subset of users
that get routed to them will just have their sqlite db file rebuilt on that machine the next time it's needed.
These sandbox db files effectively serve as a cache.

What **does** cause a problem is if a user is associated with machine A, and then gets switched over to machine
B, and then goes back to machine A. In that situation, any work done on machine A wouldn't get synced to machine B

> until the next time the user did a "sync" on machine B. Until then, they would be working form stale data. This is

Reviewer: `s/form/from/`

Formplayer in HQ
================

> Theis documentation describes how `formplayer <https://github.com/dimagi/formplayer/>`_ fits into the larger

Reviewer 2: ```suggestion
This documentation describes how `formplayer <https://github.com/dimagi/formplayer/>`_ fits into the larger
```

CommCare system, especially how it relates to CommCare HQ development. For details on building, runnning, and
contributing to formplayer, see the formplayer repository.

What Is Formplayer?
^^^^^^^^^^^^^^^^^^^

Formplayer is a Java Spring Boot server that wraps `commcare-core <https://github.com/dimagi/commcare-core>`_
and presents its main features as an HTTP API. CommCare HQ's Web Apps, App Preview, and SMS Forms features are
built on top of it:

* Web Apps is a single-page application, inlined into a CommCare HQ template, the provides a web UI backed by the formplayer API.
* App Preview is essentially the same as web apps, but embedded as a cell-phone-shaped iframe within the App Builder.
* SMS Forms serializes a form filling session over SMS in a question / answer sequence that is handled
by the main HQ process, which hits formplayer's API to send answers and get the next question.

Repository Overview
^^^^^^^^^^^^^^^^^^^

.. image:: images/formplayer_repo_overview.png

* `commcare-android <https://github.com/dimagi/commcare-android>`_: The UI layer of CommCare mobile.

> * `commcare-core <https://github.com/dimagi/commcare-core>`_: The CommCare engine, this powers both CommCare mobile and formplayer. Mobile uses the ``master`` branch, while formplayer uses the ``formplayer`` branch. The two branches have a fairly small diff. ``master`` is occasionally merged into ``formplayer``, but ``formplayer`` is never merged back into ``master``.

Reviewer 3: we do [occassionally](https://github.com/dimagi/commcare-core/pull/956) merge `formplayer` back into `master` as well

Author: thanks, fixed

* `formplayer <https://github.com/dimagi/formplayer>`_
* `commcare-hq <https://github.com/dimagi/commcare-hq>`_: HQ hosts web apps and the processes that run SMS forms.

Relevant Architectural Decisions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

While a full detailing of formplayer's architecture is beyond the scope of this document, a few architectural
decisions are particularly useful for HQ devs who are new to fromplayer to understand.

Sandboxes
+++++++++
Sharing the commcare-core code between mobile and formplayer allows us to keep CommCare Android and web apps

> essentially compatible. However, because commcare-core was first written for mobile and formplayer is a farly
> thin wrapper around it,
> paradigms that make more sense on mobile than on the web. Mobile is offline-first, so submitting

Reviewer 2: ```suggestion
essentially compatible. However, because commcare-core was first written for mobile some of the
paradigms it uses make more sense on mobile than on the web. Mobile is offline-first, so submitting
```

up newly entered data and syncing back down changes others have made are intentional steps designed not to block
someone who was unable to reach the server for hours, days, or longer. That model makes very
little sense on the always-online Web Apps, but the sync/restore process is still a core part of the working model.
There's even a "Sync" button shown to the user in the web apps UI.

Rather than always fetching the latest data from the source of truth, formplayer works off of locally synced subsets of data
like those that would be on people's phones if every user had their own phone. These "sandboxes" are stored as Sqlite DB files,
as they are on the phone. A phone typically has one db file and one user, whereas on formplayer, there
are as many db files as there are users, i.e. tens of thousands. Each file has its own slice of the data synced
down from the source of truth, but always just a little bit out of date if anyone's updated it after their last
sync.

Navigation and replaying of sessions

> ++++++++++++++++++++++++++++++++++++

Reviewer 3: nice addition 👍
